mmv1/third_party/terraform/fwprovider/framework_provider_access_token_test.go.tmpl

```diff
@@ -191,7 +191,7 @@ credentials = "%s"
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("data.google_provider_config_plugin_framework.default", "access_token", accessToken),
 					// not set as ENV not used
-					resource.TestCheckNoResourceAttr("data.google_provider_config_plugin_framework.default", "credentials"),
+					resource.TestCheckResourceAttr("data.google_provider_config_plugin_framework.default", "credentials", ""),
```

Comment:
Going from not present to "" doesn't impact users here because google_provider_config_plugin_framework isn't user-facing at all right?

Is the change in behavior coming from this addition: https://github.com/GoogleCloudPlatform/magic-modules/pull/11903/files#diff-faca97bd3809d437d4db599612e3d9eceae4102e172e3016097d0cfec117689aR206 ?

Can you help me understand fully why it was important enough to verify that it had no value before in a test, but we actually want to ensure it has an empty value now? Is it less of a "this is how the value should be" and more of a "this is how it is represented now, let's test it to make sure it doesn't change accidentally later"?

Reply:
> Going from not present to "" doesn't impact users here because google_provider_config_plugin_framework isn't user-facing at all right?

That's right, google_provider_config_plugin_framework is not user facing so we don't need to worry about that data source specifically. This data source just exposes that 'meta' Config struct so the SDK-like behaviour is fully exposed here.

**I'll look more into `google_client_config` though, as this data source exposes the provider's config to users. It is analogous to `google_provider_config_*` data sources but is user-facing. As google_client_config is PF-implemented it'll be effectively like google_provider_config_plugin_framework.**

Also, I don't believe we will expose users to this SDK-like behaviour post muxing fixes because:

1.  We stop empty strings being passed from the PF provider to PF-implemented data sources
    a. For example currently there are helper functions like [GetProjectFramework](https://github.com/GoogleCloudPlatform/magic-modules/blob/1e3b6fed59b9905d4ea14e55c9e76b0aaeb7a074/mmv1/third_party/terraform/fwresource/field_helpers.go#L23-L34) that pass provider-level default values to resources but are written to explicitly treat provider-level arguments as unset if the value is `""`.
2. Datasource/resource-level fields that can use provider-level defaults cannot be empty/unset in valid resources
    a. Provider-level defaults are used for fields that _have_ to be set for something to be valid. They're Optional in the schema but are required fields for the API.
    b. I don't believe we'll encounter similar "Going from not present to "" " changes for pre-existing resources in a users config as a valid pre-existing config would need to have provider defaults that aren't `""`
    c. `google_client_config` is the exception to this
3. The existing migrated firebase data sources : uses GetProjectFramework and requires a value for `project` so are covered by point above

Reply:
> Can you help me understand fully why it was important enough to verify that it had no value before in a test, but we actually want to ensure it has an empty value now? Is it less of a "this is how the value should be" and more of a "this is how it is represented now, let's test it to make sure it doesn't change accidentally later"?

My intention with those checks was to show how the provider behaves now, and ensure it was the same between SDK and PF. I wanted to explore how conflicting fields were handled when one is set via config and the other is set via ENVs.

Currently [credentials and access_token are set as conflicting in the provider schema](https://github.com/GoogleCloudPlatform/magic-modules/blob/8a6f281bad55b80c3e9f6d739729cb5d78646ab3/mmv1/third_party/terraform/provider/provider.go.tmpl#L33-L45), which restricts what user inputs can be supplied via config. However as the provider configuration logic pulls in values from ENVs there is a chance of those conflicting fields both being set in a way that the schema-level validation cannot detect and prevent.

In these tests I wanted to define the behaviour that if one of the conflicting fields was set in the config and the other was provided via ENV:

1. No user-facing validation logic was triggered
2. The field set in the config is the one that's used, and the conflicting field is ignored.
    a. Specifically, ignored to the point of ENVs not being used even if they're present.

Does that help?